packages/material-ui/src/useMediaQuery/useMediaQuery.js

```diff
@@ -54,7 +51,6 @@ function useMediaQuery(queryInput, options = {}) {
 
   React.useEffect(() => {
     let active = true;
-    hydrationCompleted = true;
```

Comment:
So the depth-first use effect would "complete" hydration for everyone.
This flag could be set only on top level component as long as its `useEffect` would be called after all children ones.

Reply:
@theKashey Interesting, a top-level use effect could be a good edge in case progressive hydration https://youtu.be/k-A2VfuUROg?t=993 come to React 😁.

Reply:
I am using the hack from the video for last two years (well used a bit less performant hack before)

Reply:
I start to think that we need a second top-level Provider for Material-UI (in addition to `ThemeProvider`), maybe a `ConfigProvider` that bundles different concerns so we don't end up with this API, but we can still use different context internally:

```
How
 Many
   Providers
     Until
       We
         Say
          It's   
            Enough?
```

A provider that could host:

- Number formatting https://github.com/mui-org/material-ui/pull/20656#issuecomment-617429809
- Date formatting https://next.material-ui-pickers.dev/getting-started/installation
- Date parsing https://next.material-ui-pickers.dev/getting-started/installation
- Snackbar/Dialog imperative API https://github.com/iamhosseindhv/notistack#how-to-use, https://trello.com/c/iNtH8wgp/2176-dialog-imperative-api.
- Client-side rendering only https://github.com/mui-org/material-ui/issues/21142#issuecomment-633163388
- … maybe more in the future

cc @mui-org/core

Reply:
> I am using the hack from the video for last two years (well used a bit less performant hack before)

Oh, so the useEffect at the root is a dead end? Each partially hydrated tree would need its own provider. I think that we can stick to the noSsr boolean, simpler.

Reply:
> Oh, so the useEffect at the root is a dead end?

No-no, it works just perfectly. And the additional Provider to update SSR status would be just great.